csharp/src/Ice/MetricsObserver.cs

```diff
@@ -0,0 +1,230 @@
+                return resolver(helper)?.ToString() ?? "";
+            }
+        }
+    }
+
+    internal class Observer<T> : Stopwatch, Ice.Instrumentation.IObserver where T : Metrics, new()
+    {
+        public delegate void MetricsUpdate(T m);
+
+        private List<MetricsMap<T>.Entry>? _objects;
+        private long _previousDelay = 0;
+
+        public virtual void Attach() => Start();
+
+        public virtual void Detach()
+        {
+            Stop();
+            long lifetime = _previousDelay + (long)(ElapsedTicks / (Frequency / 1000000.0));
+            foreach (MetricsMap<T>.Entry e in _objects!)
+            {
+                e.Detach(lifetime);
+            }
+        }
+
+        public virtual void Failed(string exceptionName)
+        {
+            foreach (MetricsMap<T>.Entry e in _objects!)
+            {
+                e.Failed(exceptionName);
+            }
+        }
+
+        internal void ForEach(MetricsUpdate u)
+        {
+            foreach (MetricsMap<T>.Entry e in _objects!)
+            {
+                e.Execute(u);
+            }
+        }
+
+        internal MetricsMap<T>.Entry? GetEntry(MetricsMap<T> map) => _objects.FirstOrDefault(entry => entry.Map == map);
+
+        internal ObserverImpl? GetObserver<S, ObserverImpl>(string mapName, MetricsHelper<S> helper)
+            where S : Metrics, new()
+            where ObserverImpl : Observer<S>, new()
+        {
+            List<MetricsMap<S>.Entry>? metricsObjects = null;
+            foreach (MetricsMap<T>.Entry entry in _objects!)
+            {
+                MetricsMap<S>.Entry? e = entry.GetMatching(mapName, helper);
+                if (e != null)
+                {
+                    if (metricsObjects == null)
+                    {
+                        metricsObjects = new List<MetricsMap<S>.Entry>(_objects.Count);
+                    }
+                    metricsObjects.Add(e);
+                }
+            }
+
+            if (metricsObjects == null)
+            {
+                return null;
+            }
+
+            try
+            {
+                var obsv = new ObserverImpl();
+                obsv.Init(metricsObjects, null);
+                return obsv;
+            }
+            catch (Exception ex)
+            {
+                Debug.Assert(false, ex.ToString());
+                return null;
+            }
+        }
+
+        internal void Init(List<MetricsMap<T>.Entry> objects, Observer<T>? previous)
+        {
+            _objects = objects;
+
+            if (previous == null)
+            {
+                return;
+            }
+
+            _previousDelay = previous._previousDelay + (long)(previous.ElapsedTicks / (Frequency / 1000000.0));
+            foreach (MetricsMap<T>.Entry e in previous._objects!)
+            {
+                if (!_objects.Contains(e))
+                {
+                    e.Detach(_previousDelay);
+                }
+            }
+        }
+    }
```

Comment:
We could try to get rid of the Init method which requires to specify _objects/entries as a nullable... instead of we'd have to construct the observer with a lambda rather than rely on the parameter less constructor + Init method.

Reply:
Not sure if this will work on C#, don't we need to use the default constructor because the observer can be a template argument

Reply:
I'm pretty sure we can provide and use a factory `Func<List<MetricsMap<T>.Entry>, O?, O> factory` where we need to construct the observer. I can check it out.